edx_django_utils/monitoring/internal/middleware.py

```diff
@@ -400,18 +408,71 @@ def get_log_message_and_monitor_cookies(self, request):
-        # there's indication of corruption, just log all the headers
-        # for later diagnosis.
-        if corrupt_cookie_count:
-            # .. setting_name: UNUSUAL_COOKIE_SAMPLING_PUBLIC_KEY
-            # .. setting_default: None
-            # .. setting_description: If set and there's a corrupt-looking cookie and cookie information
-            #   is being logged, then log the (encrypted) request headers as well. See log_sensitive
-            #   module for more detail on the encryption.
-            if cookie_encryption_pub_key := getattr(settings, 'UNUSUAL_COOKIE_SAMPLING_PUBLIC_KEY', None):
-                headers_plain = dict(request.headers.items())
-                headers_enc = encrypt_for_log(json.dumps(headers_plain), cookie_encryption_pub_key)
-                log_message += f" All headers, with {corrupt_cookie_count} likely corruptions: {headers_enc}"
-
         return log_message
+
+    def log_corrupt_cookie_headers(self, request, corrupt_cookie_count):
+        """
+        Log all headers when corrupt cookies are detected (if settings permit).
+
+        This log data is encrypted using the log-sensitive utility.
+
+        - Logging requires that ``UNUSUAL_COOKIE_SAMPLING_PUBLIC_KEY`` is set.
+        - Output is split across multiple lines using ``UNUSUAL_COOKIE_SAMPLING_LOG_CHUNK``.
+        """
+        # Caller should ensure this, but check here anyway.
+        if corrupt_cookie_count < 1:
+            return
+
+        # .. setting_name: UNUSUAL_COOKIE_SAMPLING_PUBLIC_KEY
+        # .. setting_default: None
+        # .. setting_description: Use this public key to encrypt and log headers when there's
+        #   a corrupted-looking cookie in the request. See log_sensitive module for more detail
+        #   on the encryption. If no key is provided, this logging is skipped. Also see
+        #   ``UNUSUAL_COOKIE_SAMPLING_LOG_CHUNK``.
+        corrupt_cookie_log_pub_key = getattr(settings, 'UNUSUAL_COOKIE_SAMPLING_PUBLIC_KEY', None)
+
+        if not corrupt_cookie_log_pub_key:
+            return
+
+        # .. setting_name: UNUSUAL_COOKIE_SAMPLING_LOG_CHUNK
+        # .. setting_default: 9000
+        # .. setting_description: If necessary, logs data in chunks of this size, splitting across
+        #   multiple log messages. This should be set with your deployment's maximum log message
+        #   size in mind: Setting it too high may result in truncated messages, which will prevent
+        #   decryption (CryptoError).
+        chunk_size = getattr(settings, 'UNUSUAL_COOKIE_SAMPLING_LOG_CHUNK', 9000)
+
+        header_data = json.dumps(dict(request.headers.items()))
+        enc_output = encrypt_for_log(header_data, corrupt_cookie_log_pub_key)
+        msg = f"All headers for request with corrupted cookies (count={corrupt_cookie_count}): {enc_output}"
+
+        for piece in split_ascii_log_message(msg, chunk_size):
+            log.info(piece)
+
+
+def split_ascii_log_message(msg, chunk_size):
```

Comment:
I may be missing something, but I was hoping if this were done, it could go into `edx_django_utils/logging/` and be used for any really long log message. Also, the setting could be something like `LONG_LOG_MESSAGE_CHUNK_SIZE`, and it would just need to be set once to match the environment.

Anyway, I don't want to block on this, but that is what I had in mind.

Reply:
Yeah. I was hoping to do that as well, but there's some fiddly stuff here around splitting on characters vs bytes when it comes to multibyte characters.

- If you split on character count, then you may exceed the line size if there are multibyte characters
- If you split on byte count, then multibyte characters like "犬" run a risk of being split between the bytes

It ended up being too much of a rabbit hole, although I'd still like to clean it up to make it generally useful—or find some library that already does this. In the meantime, it should be sufficient for this known-character-range code.